packages/workspace/src/browser/workspace-commands.ts

```diff
@@ -225,14 +223,36 @@ export class WorkspaceCommandContribution implements CommandContribution {
 
     registerCommands(registry: CommandRegistry): void {
         this.registerOpenWith(registry);
-        registry.registerCommand(WorkspaceCommands.NEW_FILE, this.newWorkspaceRootUriAwareCommandHandler({
-            execute: uri => this.getDirectory(uri).then(parent => {
+        registry.registerCommand(CommonCommands.NEW_FILE, this.newWorkspaceRootUriAwareCommandHandler({
+            execute: (uri, ...args) => this.getDirectory(uri).then(async parent => {
                 if (parent) {
-                    const parentUri = parent.resource;
                     const { fileName, fileExtension } = this.getDefaultFileConfig();
-                    const targetUri = parentUri.resolve(fileName + fileExtension);
+                    let newFileName = `${fileName}${fileExtension}`;
+                    const parentUri = parent.resource;
+
+                    const doCreateFile = async (name: string | undefined) => {
+                        if (name) {
+                            const fileUri = parentUri.resolve(name);
+                            await this.fileService.create(fileUri);
+                            this.fireCreateNewFile({ parent: parentUri, uri: fileUri });
+                            open(this.openerService, fileUri);
+                        }
+                    };
+
+                    // If a name was provided, check if it is valid, create a file and return.
+                    // Otherwise open a dialog to get the name.
+                    if (args !== undefined && args.length > 0 && args[0]) {
+                        newFileName = args[0];
+                        if (await this.validateFileName(newFileName, parent, true) === '') {
```

Comment:
We should double check whether `newFileName` is a string here because when we trigger that command from the toolbar in the File Navigator then actually the first command will be the `FileNavigatorWidget` and we will get an exception when we try to call the command from the toolbar.